control-plane/api-gateway/common/metrics.go

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package common
+
+import (
+	"strconv"
+
+	"github.com/hashicorp/consul-k8s/control-plane/api/v1alpha1"
+	"github.com/hashicorp/consul-k8s/control-plane/connect-inject/constants"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+const (
+	defaultScrapePort = 20200
+	defaultScrapePath = "/metrics"
+)
+
+type MetricsConfig struct {
+	Enabled bool
+	Path    string
+	Port    int
+}
+
+func gatewayMetricsEnabled(gateway gwv1beta1.Gateway, gcc v1alpha1.GatewayClassConfig, config HelmConfig) bool {
+	// first check our annotations, if something is there, then it means we've explicitly
+	// annotated metrics collection
+	if scrape, isSet := gateway.Annotations[constants.AnnotationEnableMetrics]; isSet {
+		return scrape == "true"
+	}
+
+	// if it's not set on the annotation, then we check to see if it's set on the GatewayClassConfig
+	if gcc.Spec.Metrics.Enabled != nil {
+		return *gcc.Spec.Metrics.Enabled
+	}
+
+	// otherwise, fallback to the global helm setting
+	return config.EnableGatewayMetrics
+}
+
+func fetchPortString(gateway gwv1beta1.Gateway, gcc v1alpha1.GatewayClassConfig, config HelmConfig) string {
+	// first check our annotations, if something is there, then it means we've explicitly
+	// annotated metrics collection
+	if portString, isSet := gateway.Annotations[constants.AnnotationPrometheusScrapePort]; isSet {
+		return portString
+	}
+
+	// if it's not set on the annotation, then we check to see if it's set on the GatewayClassConfig
+	if gcc.Spec.Metrics.Port != nil {
+		return strconv.Itoa(int(*gcc.Spec.Metrics.Port))
+	}
+
+	// otherwise, fallback to the global helm setting
+	return config.DefaultPrometheusScrapePort
+}
+
+func gatewayMetricsPort(gateway gwv1beta1.Gateway, gcc v1alpha1.GatewayClassConfig, config HelmConfig) int {
+	portString := fetchPortString(gateway, gcc, config)
+
+	port, err := strconv.Atoi(portString)
+	if err != nil {
+		// if we can't parse the port string, just use the default
+		// TODO: log an error
+		return defaultScrapePort
+	}
+
+	if port < 1024 || port > 65535 {
+		// if we requested a privileged port, use the default
+		// TODO: log an error
+		return defaultScrapePort
+	}
+
+	return port
+}
+
+func gatewayMetricsPath(gateway gwv1beta1.Gateway, gcc v1alpha1.GatewayClassConfig, config HelmConfig) string {
+	// first check our annotations, if something is there, then it means we've explicitly
+	// annotated metrics collection
+	if path, isSet := gateway.Annotations[constants.AnnotationPrometheusScrapePath]; isSet {
+		return path
+	}
+
+	// if it's not set on the annotation, then we check to see if it's set on the GatewayClassConfig
+	if gcc.Spec.Metrics.Path != nil {
+		return *gcc.Spec.Metrics.Path
+	}
+
+	// otherwise, fallback to the global helm setting
+	return config.DefaultPrometheusScrapePath
```

Comment:
Should we be doing any validation on the path here?

Reply:
Ideally, yes, but from what I can tell we don't really have solid path validation anywhere else in our implementation (unless you can point me to a prior art?). Was thinking we'd just keep things equivalent for now, but we could see if we could `url.Parse` or regex match the url string if we wanted to? I believe that if this is invalid you'll just get a dataplane process that fails to start up envoy.